### PR TITLE
Bug Fix to prevent repeated writes to metadata.json

### PR DESCRIPTION
**Pull request by anirudhacharya**
### Description of changes:
During a Sagemaker training job, there was an issue of the `metadata.json` file in the `checkpoints_dir` being re-written every second.

This probably happened because we use `json.dumps` [here](https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/core/state_store.py#L104) to write to the `states_file` every time the `last_modified_time` of the `checkpoint_dir` is greater than `last_Saved time`( [checked here](https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/core/state_store.py#L75)). 

`json.dumps` may not guarantee that file was actually written to disk. The file could be in cache and flushed to disk asynchronously a second later. In such a case the `last_modified_time` of the file will always be more than `last_Saved time`. 

To fix it I have filtered out the `states_file` while determining if the `checkpoint_dir` has been updated( [here](https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/core/state_store.py#L82)).

#### Style and formatting:

I have run `pre-commit install` to ensure that auto-formatting happens with every commit.

#### Issue number, if available

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by anirudhacharya** [comment]
There are a few issues with this fix that I have made. 
1. By modifying `is_checkpoint_updated` we have changed the nature of the function. By filtering out the `metadata.json` file we have  made the function inaccurate. So now even if the `checkpoint_dir` is modified `is_checkpoint_updated` function can still return `False`.

For example, the checkpoint_dir is like this -
```bash
checkpoint_dir
 |-- sub-dir1
 |-- sub-dir2
 |-- sub-dir3
 |-- metadata.json
```
If `metadata.json` is modified then it technically means that `checkpoint_dir` is also modified, so `is_checkpoint_updated` should return `True`. But the current modification to the function will not do that.

One way to fix the bug without modifying the behavior of the `is_checkpoint_updated` function might be to ensure that the write call to write the `states_file` is a synchronous call( [here](https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/core/state_store.py#L104))

@Vikas-kum kindly review this.

File: smdebug/core/state_store.py [hunk 1]
```diff
@@ -79,7 +79,9 @@ def is_checkpoint_updated(self):
         """
         if self._checkpoint_dir is not None:
             checkpoint_timestamp = max(
-                os.path.getmtime(child) for child, _, _ in os.walk(self._checkpoint_dir)
+                os.path.getmtime(child)
+                for child, _, _ in os.walk(self._checkpoint_dir)
+                if child != os.path.dirname(self._states_file)
```
Comment (Vikas-kum) [inline smdebug/core/state_store.py:84]: why dirname? You need to check for exact filename and filter out meradata.json 
Also write a test case for this.

File: smdebug/core/state_store.py [hunk 2]
```diff
@@ -74,14 +74,24 @@ def _read_states_file(self):
...
+                for child, _, _ in os.walk(self._checkpoint_dir)
+                if child != self._checkpoint_dir
+            ]
+            _, _, files = next(os.walk(self._checkpoint_dir))
+            timestamps += [
+                os.path.getmtime(os.path.join(self._checkpoint_dir, file))
+                for file in files
+                if file != METADATA_FILENAME
+            ]
+            if not timestamps:
```
Comment (Vikas-kum) [inline smdebug/core/state_store.py:92]: I think we need a simple way to recursively get all the files.  
import os
>>> for child, r, c in os.walk("/tmp/newlogsRunTest/"):
...   for f in c:
...     if f != "haha":
...       print(child,"/",f)

Also see here- 
https://stackoverflow.com/questions/2186525/how-to-use-glob-to-find-files-recursively

File: smdebug/core/state_store.py [hunk 3]
```diff
@@ -74,14 +74,24 @@ def _read_states_file(self):
...
+                return False
+            if max(timestamps) > self._checkpoint_update_timestamp:
```
Comment (Vikas-kum) [inline smdebug/core/state_store.py:94]: before returning we should also add logger debug and dump files considered anf their gmtime

File: tests/core/test_state_store.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+
+
+def test_is_checkpoint_updated():
+    s1 = StateStore()
+    s1._checkpoint_update_timestamp = time.time()
+    s1._checkpoint_dir = setup_test1()
+    # the checkpoint update time is greater than _checkpoint_update_timestamp. is_checkpoint_updated should return true.
+    assert s1.is_checkpoint_updated()
+
+    s1._checkpoint_update_timestamp = time.time()
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:49]: why this i set in test case. 

I think you require test case like:
test1:

State_store - check_updated() ?  returns False
create dir/files 
State_store - check_updated() ?  returns True 
update_state() 
state_store -> check_updated ? reutrns False 

update one of existing file : 
state_store -> check_updated ? returns True 
update_state()
Create _one_ new file 
state_store -> check_updated ? returns True

**Comment by anirudhacharya** [comment]
> Also, please look at build failures.

this test `test_is_checkpoint_updated` which I added in this PR, failed. Surprising, since all these tests passed locally. I will see, what might have gone wrong.

**Comment by anirudhacharya** [comment]
Three tf2 job will pass after I rebase on top of PR #203

File: config/buildspec_zero_code_change.yml [hunk 5]
```diff
@@ -31,8 +31,8 @@ phases:
 
   build:
     commands:
-      - cd $CODEBUILD_SRC_DIR  && chmod +x config/tests.sh && PYTHONPATH=. ./config/tests.sh && cd ..
-      - cd $CODEBUILD_SRC_DIR_RULES && chmod +x config/tests.sh && PYTHONPATH=. ./config/tests.sh  && cd ..
+      - cd $CODEBUILD_SRC_DIR  && chmod +x config/tests.sh && PYTHONPATH=. && ./config/tests.sh && cd ..
```
Comment (Vikas-kum) [inline config/buildspec_zero_code_change.yml:34]: why did we need to change this?

File: tests/core/test_state_store.py [hunk 6]
```diff
@@ -54,6 +55,7 @@ def test_is_checkpoint_updated():
     assert s2.is_checkpoint_updated() is False
 
     s2.update_state("test-state1")
+    time.sleep(1)
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:58]: why is this needed?

File: tests/core/test_state_store.py [hunk 7]
```diff
@@ -64,6 +66,7 @@ def test_is_checkpoint_updated():
     assert s2.is_checkpoint_updated()
 
     s2.update_state("test-state2")
+    time.sleep(1)
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:69]: why is this needed?

File: tests/core/test_state_store.py [hunk 8]
```diff
@@ -75,10 +78,12 @@ def test_is_checkpoint_updated():
     assert s2.is_checkpoint_updated()
 
     s2.update_state("test-state3")
+    time.sleep(1)
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:81]: same here?

File: tests/core/test_state_store.py (hunk 6 above)
Comment (anirudhacharya) [reply tests/core/test_state_store.py:58]: because `update_state` is an async call. 
Even though I am updating `/subdir1/checkpoint_test1.txt` in line 63 after this call, the actual `update_state` happens after `checkpoint_test1.txt` is written. So while we expect line 66 to be `True`, it will return `False` without the 1 second delay.

I tested this out as follows -
```
tests/core/test_state_store.py::test_is_checkpoint_updated 
last_timestamp: 1586394007.8299778 and timestamp of chkpnt file: 1586394007.8291862
```
`last_timestamp` is when `update_state` writes to file to disk(line 57), and `timestamp of chkpnt file` is when the checkpoint file is written to disk(line64).

File: tests/core/test_state_store.py (hunk 7 above)
Comment (anirudhacharya) [reply tests/core/test_state_store.py:69]: same as above

File: tests/core/test_state_store.py (hunk 8 above)
Comment (anirudhacharya) [reply tests/core/test_state_store.py:81]: same as above.

File: tests/core/test_state_store.py (hunk 6 above)
Comment (anirudhacharya) [reply tests/core/test_state_store.py:58]: this is the failed build link - https://tiny.amazon.com/114zqzab7/IsenLink

File: config/buildspec_zero_code_change.yml (hunk 5 above)
Comment (anirudhacharya) [reply config/buildspec_zero_code_change.yml:34]: we don't need this. it does not cause any failure. But I can remove it. I am not sure how this change got into this PR.

Comment (Vikas-kum) [reply config/buildspec_zero_code_change.yml:34]: please remove if not needed.

File: tests/core/test_state_store.py (hunk 6 above)
Comment (Vikas-kum) [reply tests/core/test_state_store.py:58]: why would it fail? you say update_state() , a new json file would be written. 
Since there is no new checkpoint file created, irrespective of whether update_state wrote the file or not youshould get False as result of is_checkpoint_updated.
Am I missing something?

File: tests/core/test_state_store.py [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    assert s2.is_checkpoint_updated() is False
+
+    s2.update_state("test-state1")
+    time.sleep(1)
+    # checkpoints_dir still has only metadata.json. It should return false.
+    assert s2.is_checkpoint_updated() is False
+
+    os.mkdir(s2._checkpoint_dir + "/subdir1")
+    with open(s2._checkpoint_dir + "/subdir1/checkpoint_test1.txt", "w") as f:
+        f.write("checkpoint-test-string-1")
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:64]: sync here . f.flush() to make sure that next line is certain to return true.

File: tests/core/test_state_store.py (hunk 7 above)
Comment (Vikas-kum) [reply tests/core/test_state_store.py:69]: As I said earlier, I believe you shouldn't need this and below

File: tests/core/test_state_store.py [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+    # A checkpoint file has been updated. The checkpoint update time is greater than _checkpoint_update_timestamp.
+    # is_checkpoint_updated should return true.
+    assert s2.is_checkpoint_updated()
+
+    s2.update_state("test-state3")
+    time.sleep(1)
+    os.mkdir(s2._checkpoint_dir + "/subdir2")
+    with open(s2._checkpoint_dir + "/subdir2/checkpoint_test2.txt", "w") as f:
+        f.write("checkpoint-test-string-3")
+    # A new checkpoint file has been created. The checkpoint update time is greater than _checkpoint_update_timestamp.
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:85]: f.flush() to make sure that file is written. Also, make sure that flush is right call.

File: tests/core/test_state_store.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    mock_metadata = [
+        {
+            "training-run": "",
+            "latest-global-step-saved": "",
+            "latest-global-step-seen": "",
+            "latest-mode-step": "",
+        }
+    ]
+    with open(dir_path + "/metadata.json", "w") as f:
+        json.dump(mock_metadata, f)
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:36]: f.flush()

File: tests/core/test_state_store.py [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+    assert s2.is_checkpoint_updated()
+
+    s2.update_state("test-state2")
+    time.sleep(1)
+    # the state_file has been updated. The lastest checkpoint update time is lesser than _checkpoint_update_timestamp.
+    # is_checkpoint_updated should return false.
+    assert s2.is_checkpoint_updated() is False
+
+    with open(s2._checkpoint_dir + "/subdir1/checkpoint_test1.txt", "a") as f:
+        f.write("checkpoint-test-string-2")
```
Comment (Vikas-kum) [inline tests/core/test_state_store.py:75]: flush.

File: tests/core/test_state_store.py (hunk 6 above)
Comment (anirudhacharya) [reply tests/core/test_state_store.py:58]: yes you should get False in line 60 and True in line 66. the `time.sleep` is to ensure a True value in line 66.

File: tests/core/test_state_store.py (hunk 9 above)
Comment (anirudhacharya) [reply tests/core/test_state_store.py:64]: flush here might not make any difference. 

We could probably remove the sleep if we did a `fileObject.flush` and `os.fsync(f)` here - [state_store.py#L104](https://github.com/awslabs/sagemaker-debugger/blob/master/smdebug/core/state_store.py#L104)

Comment (Vikas-kum) [reply tests/core/test_state_store.py:64]: yeah, this is what I suggested earlier, please remove the sleep from code and use flush(method which does fsync) . we dont neeed to do fsync in code, but in test

File: tests/core/test_state_store.py (hunk 10 above)
Comment (Vikas-kum) [reply tests/core/test_state_store.py:85]: use flush + fsync

**Comment by Vikas-kum** [comment]
https://unix.stackexchange.com/questions/11599/determine-file-system-timestamp-precision#12756
file system time precision is causing this probably. Ok to put sleep before when you create new files ( Write down reason in comments that https://unix.stackexchange.com/questions/11599/determine-file-system-timestamp-precision#12756) 
Put sleep of 2 seconds to be safe.

**Comment by anirudhacharya** [comment]
> https://unix.stackexchange.com/questions/11599/determine-file-system-timestamp-precision#12756
> file system time precision is causing this probably.
> what if same thing happens in prodcution, last_update_timestamp > gmtime and it will say that file was not updated and miss the new checkpoint file if timediff between last_updatetimestamp and gmtime is ~1 sec say

there is a lag between the time reported by python's time `module` and system time reported by `os.path.getmtime()`. Here is a code snippet that demonstrates that 

```python
import time
import os
import json

print(f"t0: {time.time()}")

with open("status.json", 'w') as f:
    json.dump("status", f)

t1 = os.path.getmtime("status.json")
print(f"t1: {t1}")

print(f"t2: {time.time()}")

print(f"t3: {time.time()}")

with open("test.txt", 'w') as f:
    f.write("test")
t4 = os.path.getmtime("test.txt")
print(f"t4: {t4}")
print(f"t5: {time.time()}")
```
Output:
```bash
t0: 1586555897.3936992
t1: 1586555897.388582
t2: 1586555897.3938596
t3: 1586555897.393874
t4: 1586555897.388582
t5: 1586555897.3939729
```

**Comment by Vikas-kum** [comment]
fixed in PR:https://github.com/awslabs/sagemaker-debugger/pull/200
